modules/core/src/main/scala/org/tessellation/domain/snapshot/programs/Download.scala

```diff
@@ -0,0 +1,60 @@
+package org.tessellation.domain.snapshot.programs
+
+import cats.effect.Async
+import cats.syntax.applicativeError._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.option._
+
+import org.tessellation.domain.snapshot.GlobalSnapshotStorage
+import org.tessellation.http.p2p.clients.GlobalSnapshotClient
+import org.tessellation.infrastructure.snapshot._
+import org.tessellation.schema.node.NodeState
+import org.tessellation.sdk.domain.cluster.storage.ClusterStorage
+import org.tessellation.sdk.domain.node.NodeStorage
+import org.tessellation.sdk.infrastructure.consensus.Consensus
+
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+object Download {
+
+  def make[F[_]: Async](
+    nodeStorage: NodeStorage[F],
+    clusterStorage: ClusterStorage[F],
+    globalSnapshotClient: GlobalSnapshotClient[F],
+    globalSnapshotStorage: GlobalSnapshotStorage[F],
+    consensus: Consensus[F, GlobalSnapshotKey, GlobalSnapshotArtifact]
+  ): Download[F] = new Download(nodeStorage, clusterStorage, globalSnapshotClient, globalSnapshotStorage, consensus) {}
+}
+
+sealed abstract class Download[F[_]: Async] private (
+  nodeStorage: NodeStorage[F],
+  clusterStorage: ClusterStorage[F],
+  globalSnapshotClient: GlobalSnapshotClient[F],
+  globalSnapshotStorage: GlobalSnapshotStorage[F],
+  consensus: Consensus[F, GlobalSnapshotKey, GlobalSnapshotArtifact]
+) {
+  private def logger = Slf4jLogger.getLogger[F]
+
+  def download(): F[Unit] =
+    nodeStorage.tryModifyState(NodeState.WaitingForDownload, NodeState.DownloadInProgress, NodeState.Ready) {
+      clusterStorage.getPeers
+        .map(_.headOption)
+        .flatMap {
+          case None =>
+            (new Throwable(s"Unexpected state during download. No peer found, but node should be already connected."))
```

Comment:
It's a forbidden state that will never throw, because it happens only after successful joining.
Option is because of implementation, that's why a generic error string, but anyway it will never happen.